src/cmd/ksh93/sh/lex.c

```diff
@@ -340,11 +354,7 @@ int sh_lex(Lex_t *lp) {
     lp->typed = 0;
 
     while (1) {
-        // Skip over characters in the current state.
-        state = sh_lexstates[mode];
-        while ((n = STATE(state, c)) == 0) {
-            ;  // empty loop
-        }
+        read_to_next_state_transition(mode, &state, &n, &c);
```

Comment:
I don't think the `state = sh_lexstates[mode];` statement belongs in this new function as that assignment is orthogonal to the purpose of the while loop you moved into that function. And `n` should be passed in by value and returned by the function.

Note that there are only two places that use the `STATE` macro. It is both badly named and the wrong abstraction. Mostly because the author chose to implement it as a macro rather than a function. The assignment to `c` should be done separate from the state table lookup, for example. And it should probably be named something like `NEXT_TOKEN_STATE`. Which assumes we kept it as a macro which we probably shouldn't.